src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala

```diff
@@ -22,6 +22,31 @@ final class BCodeAsmCommon[G <: Global](val global: G) {
   }
 
   /**
+   * True for classes generated by the Scala compiler that are considered top-level in terms of
+   * the InnerClass / EnclosingMethod classfile attributes. See comment in BTypes.
+   */
+  def considerAsTopLevelImplementationArtifact(classSym: Symbol) = {
+    classSym.isImplClass || classSym.isSpecialized
+  }
+
+  /**
+   * Cache the value of delambdafy == "inline" for each run. We need to query this value many
+   * times, so caching makes sense.
+   */
+  object delambdafyInline {
+    private var runId = -1
+    private var value = false
+
+    def apply(): Boolean = {
+      if (runId != global.currentRunId) {
+        runId = global.currentRunId
+        value = settings.Ydelambdafy.value == "inline"
+      }
+      value
+    }
+  }
```

Comment:
No need to change this for this PR, but let's think about whether it would make sense to move this (and maybe other similar code) into a new class `RunSettings`, accessible via `currentRun.settings._`.